src/python/grpcio_tests/tests/observability/_open_telemetry_observability_test.py

```diff
@@ -0,0 +1,232 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import sys
+import time
+from typing import Any, Dict, Optional, Set
+import unittest
+
+import grpc_observability
+from grpc_observability import _open_telemetry_measures
+from grpc_observability._open_telemetry_plugin import GRPC_METHOD_LABEL
+from grpc_observability._open_telemetry_plugin import GRPC_OTHER_LABEL_VALUE
+from grpc_observability._open_telemetry_plugin import GRPC_TARGET_LABEL
+from opentelemetry.sdk.metrics import MeterProvider
+from opentelemetry.sdk.metrics.export import AggregationTemporality
+from opentelemetry.sdk.metrics.export import MetricExportResult
+from opentelemetry.sdk.metrics.export import MetricExporter
+from opentelemetry.sdk.metrics.export import MetricsData
+from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
+
+from tests.observability import _test_server
+
+logger = logging.getLogger(__name__)
+
+STREAM_LENGTH = 5
+OTEL_EXPORT_INTERVAL_S = 0.5
+
+
+class OTelMetricExporter(MetricExporter):
+    """Implementation of :class:`MetricExporter` that export metrics to the
+    provided metric_list.
+    """
+
+    def __init__(
+        self,
+        all_metrics: Dict[str, Any],
+        preferred_temporality: Dict[type, AggregationTemporality] = None,
+        preferred_aggregation: Dict[
+            type, "opentelemetry.sdk.metrics.view.Aggregation"
+        ] = None,
+    ):
+        super().__init__(
+            preferred_temporality=preferred_temporality,
+            preferred_aggregation=preferred_aggregation,
+        )
+        self.all_metrics = all_metrics
+
+    def export(
+        self,
+        metrics_data: MetricsData,
+        timeout_millis: float = 10_000,
+        **kwargs,
+    ) -> MetricExportResult:
+        self.record_metric(metrics_data)
+        return MetricExportResult.SUCCESS
+
+    def shutdown(self, timeout_millis: float = 30_000, **kwargs) -> None:
+        pass
+
+    def force_flush(self, timeout_millis: float = 10_000) -> bool:
+        return True
+
+    def record_metric(self, metrics_data: MetricsData) -> None:
+        for resource_metric in metrics_data.resource_metrics:
+            for scope_metric in resource_metric.scope_metrics:
+                for metric in scope_metric.metrics:
+                    self.all_metrics["name"].add(metric.name)
+                    for data_point in metric.data.data_points:
+                        for key, value in data_point.attributes.items():
```

Comment:
Nit: Could use `defaultdict(set)` for `self.all_metrics["label"]` and then just do `self.all_metrics["label"][key].add(value)`